// vscode-js-debug-browsers/dist/darwinChrome.js
"use strict";
/*---------------------------------------------------------
 * Copyright (C) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.DarwinChromeBrowserFinder = void 0;
const util_1 = require("./util");
const darwinFinderBase_1 = require("./darwinFinderBase");
/**
 * Finds the Chrome browser on OS X.
 */
class DarwinChromeBrowserFinder extends darwinFinderBase_1.DarwinFinderBase {
    constructor() {
        super(...arguments);
        /**
         * @override
         */
        this.wellKnownPaths = [
            {
                path: '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome',
                quality: "stable" /* Stable */,
            },
            {
                path: '/Applications/Google Chrome Canary.app/Contents/MacOS/Google Chrome Canary',
                quality: "canary" /* Canary */,
            },
            {
                path: '/Applications/Google Chrome Beta.app/Contents/MacOS/Google Chrome Beta',
                quality: "beta" /* Beta */,
            },
            {
                path: '/Applications/Google Chrome Dev.app/Contents/MacOS/Google Chrome Dev',
                quality: "dev" /* Dev */,
            },
        ];
    }
    /**
     * @override
     */
    async findAll() {
        const suffixes = ['/Contents/MacOS/Google Chrome Canary', '/Contents/MacOS/Google Chrome'];
        const defaultPaths = [
            '/Applications/Google Chrome.app',
            '/Applications/Google Chrome Canary.app',
        ];
        const installations = await this.findLaunchRegisteredApps('google chrome\\( canary\\)\\?.app', defaultPaths, suffixes);
        return util_1.sort(installations, this.createPriorities([
            {
                name: 'Chrome.app',
                weight: 0,
                quality: "stable" /* Stable */,
            },
            {
                name: 'Chrome Canary.app',
                weight: 1,
                quality: "canary" /* Canary */,
            },
        ]));
    }
    getPreferredPath() {
        return this.env.CHROME_PATH;
    }
}
exports.DarwinChromeBrowserFinder = DarwinChromeBrowserFinder;
