// vscode-js-debug-browsers/dist/windowsEdge.js
"use strict";
/*---------------------------------------------------------
 * Copyright (C) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.WindowsEdgeBrowserFinder = void 0;
const path_1 = require("path");
const util_1 = require("./util");
/**
 * Finds the Chrome browser on Windows.
 */
class WindowsEdgeBrowserFinder {
    constructor(env, fs) {
        this.env = env;
        this.fs = fs;
    }
    async findWhere(predicate) {
        return (await this.findAll()).find(predicate);
    }
    async findAll() {
        const suffixes = [
            {
                name: `${path_1.sep}Microsoft${path_1.sep}Edge SxS${path_1.sep}Application${path_1.sep}msedge.exe`,
                type: "canary" /* Canary */,
            },
            {
                name: `${path_1.sep}Microsoft${path_1.sep}Edge Dev${path_1.sep}Application${path_1.sep}msedge.exe`,
                type: "dev" /* Dev */,
            },
            {
                name: `${path_1.sep}Microsoft${path_1.sep}Edge Beta${path_1.sep}Application${path_1.sep}msedge.exe`,
                type: "beta" /* Beta */,
            },
            {
                name: `${path_1.sep}Microsoft${path_1.sep}Edge${path_1.sep}Application${path_1.sep}msedge.exe`,
                type: "stable" /* Stable */,
            },
        ];
        const installations = await util_1.findWindowsCandidates(this.env, this.fs, suffixes);
        const customEdgePath = await util_1.preferredEdgePath(this.fs, this.env);
        if (customEdgePath) {
            installations.unshift({ path: customEdgePath, quality: "custom" /* Custom */ });
        }
        return installations;
    }
}
exports.WindowsEdgeBrowserFinder = WindowsEdgeBrowserFinder;
